pyrogram/client/client.py

```diff
@@ -580,37 +572,29 @@ def download_worker(self):
                             progress=progress
                         )
 
-                if tmp_file_name is None:
-                    return None
+                if temp_file_path:
+                    final_file_path = os.path.join(directory, file_name)
 
-                if file_name is not None:
-                    path[0] = os.path.join(download_directory, file_name)
-
-                try:
-                    os.remove(os.path.join(download_directory, file_name))
-                except OSError:
-                    pass
-                finally:
                     try:
-                        if download_directory:
-                            os.makedirs(download_directory, exist_ok=True)
-                        else:
-                            os.makedirs(os.path.dirname(file_name), exist_ok=True)
+                        os.remove(final_file_path)
+                    except OSError:
+                        pass
 
-                        # avoid errors moving between drives/partitions etc.
-                        shutil.move(tmp_file_name, os.path.join(download_directory, file_name))
-                    except OSError as e:
-                        log.error(e, exc_info=True)
+                    os.renames(temp_file_path, final_file_path)
```

Comment:
Just one thing with this.

In my case, my projects live on another disk drive. This lead me to discover that **os.renames** will error out if you try to rename to a location on another drive.

Since on Windows where I am currently testing things, the temp directory always exists on **C:**, unless you have 

When it goes to move the temporary file from **C:** to **J:** where the tests working directory is, it will be unable to do this and cause an exception.

Hence the use of **os.makedirs** and **shutil.move** in my code instead.  

**shutil.move** is able to move files from one disk / partition to another without this error.

https://docs.python.org/3/library/os.html#os.rename

According to the **os.rename** doc, this can also be the case on some Unix systems.

https://docs.python.org/3/library/shutil.html#shutil.move

**shutil.move** uses **os.rename** if **dst** and **src** are on the same disk, but you must make the directory tree that the file is going to be moved to before you move it.

Reply:
This sort of error occurs running the second test snippet with the use of **os.rename**, when the temp file exists on a different disk that the destination.

https://i.imgur.com/8Pg91JL.png

Reply:
Something along the lines of this at the **if** branch on line 575 fixes for me:

``` python
if temp_file_path:
    final_file_path = os.path.join(directory, file_name)

    final_file_dir = os.path.dirname(final_file_path)

    if final_file_dir:
        os.makedirs(final_file_dir, exist_ok=True)

    shutil.move(temp_file_path, final_file_path)
```

**shutil.move** overwrites **final_file_path**, so there should not be a need to delete any existing file first.

This should work unless there is something about **shutil.move** that you want to avoid that I am missing.

Reply:
Very good, but this would work the same:

``` python
if temp_file_path:
    final_file_path = os.path.join(directory, file_name)
    os.makedirs(directory, exist_ok=True)
    shutil.move(temp_file_path, final_file_path)
```

The directory path is already computed.

A thing we might also want to improve is the path looking, 'cause I'm getting paths like `asd/asddd\sticker.webp`

Reply:
I am unsure where the mixed path separator may be coming from at the moment.

But as long as **directory** is never an empty string the above should do.

Because ```os.makedirs("", exist_ok=True)``` with throw **FileNotFoundError**.

Reply:
> A thing we might also want to improve is the path looking, 'cause I'm getting paths like asd/asddd\sticker.webp

What could be done to make the returned path consistent is call **os.path.abspath** on it when putting the result in **path[0]**.  That way it always has the correct path separators for your platform.

Reply:
> What could be done to make the returned path consistent is call os.path.abspath on it when putting the result in path[0]. That way it always has the correct path separators for your platform.

Good, an absolute path looks even better than a relative one, but we need both **os.path.abspath** and the substitution because I'm getting `/home/dan/asd/asdads\asdafnaopinf\\\\asfopo/aosfnia\\aasdd` in Linux with os.path.abspath only.

> But as long as directory is never an empty string the above should do.

There's a check, if the directory is empty it will default to "downloads", so yeah, it can never be empty.